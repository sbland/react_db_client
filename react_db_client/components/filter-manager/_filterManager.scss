$unit: 1rem;//var(--unit, 1rem);
$white: white;
$defaultBoxShadow: var(--default-box-shadow, 1px 1px);
$primaryColour: var(--primary-color, #888);
$font-size-1: var(--font-size-1, 1rem);

.filterPanel {
  // position: relative;
}

.filterPanel_panel {
  /* Turned all these off as panel now just sits in section */
  // position: absolute;
  // background: $white-off;
  // padding: $unit;
  // border: 1px $primaryColour solid;
  // border-radius: 2px;
  // box-shadow: $defaultBoxShadow;
  // min-width: ($unit * 30);

  &.floating {
    position: absolute;
    padding: $unit;
    background: $white;
    box-shadow: $defaultBoxShadow;

    min-width: 30%;
    max-width: 100%;
    width: 100%;
    border: 1px $primaryColour solid;
    z-index: 100;
  }
}

.filterPanel_filterList {
  list-style: none;
}

.filterPanel_filterItem {
  margin-bottom: $unit;
  justify-content: stretch;
  display: flex;
  text-overflow: ellipsis;
  overflow: hidden;
  align-content: stretch;
  align-items: bottom;
  height: calc($unit * 2);
  vertical-align: middle;
  margin-right: -$unit;

  * {
    font-size: $font-size-1;
    line-height: ($unit * 2);
    height: ($unit * 2);
    margin-right: $unit;
  }

  :nth-child(2) {
    // flex-basis: 40%;
    width: 30%;
    white-space: nowrap;
    text-align: right;
    text-overflow: ellipsis;
    overflow: hidden;
  }

  :nth-child(3) {
    width: 20%;
    height: 100%;
  }

  :nth-child(4) {
    // flex-basis: 40%;
    width: 20%;
    flex-grow: 1;
    height: 100%;
  }
}
