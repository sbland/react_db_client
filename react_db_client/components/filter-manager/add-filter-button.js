import React from 'react';
import PropTypes from 'prop-types';


import {
  FilterObjectClass,
} from '@samnbuk/react_db_client.constants.client-types';

export const AddFilterButton = ({
  fieldsData,
  returnNewFilter,
  customFilters,
}) => {
  const handleNewFilter = () => {
    // Add a new filter and use the first field as initial field
    const initialFieldData = Object.values(fieldsData)[0];
    const { uid, field, type, label } = initialFieldData;
    const newFilter = new FilterObjectClass({
      field: field || uid,
      type,
      label,
      filterOptionId: uid,
      isCustomType: Object.keys(customFilters).indexOf(type) >= 0,
    });
    returnNewFilter(newFilter);
  };

  return (
    <button
      type="button"
      className="button-two addFilterBtn"
      onClick={handleNewFilter}
    >
      Add Filter
    </button>
  );
};

AddFilterButton.propTypes = {
  returnNewFilter: PropTypes.func.isRequired,
  fieldsData: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
    })
  ).isRequired,
  customFilters: PropTypes.objectOf(PropTypes.func),
};

AddFilterButton.defaultProps = {
  customFilters: {},
};
