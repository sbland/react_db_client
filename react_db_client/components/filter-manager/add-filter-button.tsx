import React from 'react';
import PropTypes from 'prop-types';

import { FilterObjectClass, FilterOption } from '@react_db_client/constants.client-types';
import { TFilterFunc } from './lib';

export interface IAddFilterButtonProps {
  fieldsData: { [key: string]: FilterOption<any, boolean> };
  returnNewFilter: <Vtype, IsCustomType extends boolean>(
    newFilter: FilterObjectClass<Vtype, IsCustomType>
  ) => void;
  customFilters?: { [key: string]: TFilterFunc };
}

export const AddFilterButton: React.FC<IAddFilterButtonProps> = ({
  fieldsData,
  returnNewFilter,
  customFilters = {},
}) => {
  const handleNewFilter = () => {
    // Add a new filter and use the first field as initial field
    const initialFieldData = Object.values(fieldsData)[0];
    const { uid, field, type, label } = initialFieldData;
    const isCustomFieldType = Object.keys(customFilters).indexOf(type) >= 0;

    const newFilter = new FilterObjectClass({
      field: field || uid,
      type,
      label,
      filterOptionId: uid,
      isCustomType: isCustomFieldType,
    });
    returnNewFilter(newFilter);
  };

  return (
    <button type="button" className="button-two addFilterBtn" onClick={handleNewFilter}>
      Add Filter
    </button>
  );
};

AddFilterButton.propTypes = {
  returnNewFilter: PropTypes.func.isRequired,
  fieldsData: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
    })
  ).isRequired,
  customFilters: PropTypes.objectOf(PropTypes.func),
};

AddFilterButton.defaultProps = {
  customFilters: {},
};
