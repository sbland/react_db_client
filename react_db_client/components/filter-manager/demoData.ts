import {
  comparisons,
  filterTypes,
  FilterObjectClass,
  EFilterType,
} from '@react_db_client/constants.client-types';
import { IField } from './lib';

export const demoFilterString = new FilterObjectClass({
  uid: 'demoFilterString',
  field: 'name',
  value: 'Foo',
  operator: comparisons.contains,
  type: filterTypes.text,
});
export const demoFilterNumber = new FilterObjectClass({
  uid: 'demoFilterNumber',
  field: 'count',
  value: 0,
  operator: comparisons.greaterThan,
  type: filterTypes.number,
});
// export const demoFilterExpression = new FilterObjectClass({
//   uid: 'demoFilterExpression',
//   field: 'expression',
//   value: 0,
//   operator: '>',
// });

export const demoFilterSelect = new FilterObjectClass({
  uid: 'demoFilterExpression',
  field: 'select',
  value: 'a',
  operator: comparisons.contains,
  type: filterTypes.select,
});

export const demoFieldsData = {
  name: {
    uid: 'name',
    label: 'Name',
    type: filterTypes.text,
  },
  count: {
    uid: 'count',
    label: 'Count',
    type: filterTypes.number,
  },
  filterA: {
    uid: 'filterA',
    label: 'filter A',
    type: filterTypes.text,
  },
  filterB: {
    uid: 'filterB',
    label: 'filter B',
    type: filterTypes.text,
  },
  select: {
    uid: 'select',
    label: 'Select',
    type: filterTypes.select,
    options: [
      { uid: 'a', label: 'A' },
      { uid: 'b', label: 'B' },
    ],
  },
  filterEmbedded: {
    uid: 'filterEmbedded',
    label: 'Embedded Filter',
    type: filterTypes.embedded,
    filters: {
      filterC: {
        uid: 'filterC',
        label: 'filter C',
        type: filterTypes.text,
      },
      filterD: {
        uid: 'filterD',
        label: 'filter D',
        type: filterTypes.text,
      },
    },
  },
};

export const demoFiltersData = [
  demoFilterString,
  demoFilterNumber,
  // demoFilterExpression,
  demoFilterSelect,
];

export const exampleOptions = [
  { uid: 'foo', label: 'Foo' },
  { uid: 'bar', label: 'Bar' },
];

const selectionTypes = [EFilterType.select, EFilterType.selectMulti];

export const allTypeFieldsData = Object.keys(filterTypes)
  .map((key) => ({
    uid: key,
    label: key,
    type: key,
    options: selectionTypes.indexOf(key as EFilterType) !== -1 ? exampleOptions : undefined,
  }))
  .reduce((acc, v) => ({ ...acc, [v.uid]: v }), {}) as Record<EFilterType, IField>;

export const allTypeFilters: FilterObjectClass[] = Object.values(allTypeFieldsData).map(
  (field) =>
    new FilterObjectClass({
      ...field,
      field: field.uid,
      value: undefined,
      // expression: null,
      type: field.type,
    })
);
