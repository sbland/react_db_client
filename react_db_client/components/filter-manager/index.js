export { FilterPanel } from './filter-manager';
