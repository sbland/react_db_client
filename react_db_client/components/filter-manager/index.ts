export { FilterPanel } from './filter-manager';
export type { IFilterPanelProps } from './filter-manager';
export * as FilterTypes from './FilterTypes';
export * from './useGetFilterComponents';
export * from './useManageFilters';
export * from './lib';
