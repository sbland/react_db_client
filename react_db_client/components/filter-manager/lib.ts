import { FilterObjectClass, Uid } from '@react_db_client/constants.client-types';
import { FilterOption } from '@react_db_client/constants.client-types';

export interface ILabeled {
  uid: string;
  label: string;
}

export type TFieldId = Uid;
// export interface IFieldCommon {
//   uid: TFieldId;
//   label: string;
//   field?: string;
//   noFilter?: boolean;
//   validComparisons?: EComparisons[];
//   options?: ILabeled[];
//   multiple?: boolean;
// }
// export interface IField extends IFieldCommon {
//   type: EFilterType;
//   validComparisons?: EComparisons[];
//   isCustomFieldType?: false;
//   filters?: { [k: string]: IField };
// }

// export interface ICustomField extends IFieldCommon {
//   validComparisons: EComparisons[];
//   type: string;
//   isCustomFieldType: true;
// }

export type FilterId = string | number;

export interface IFilterComponentProps<VType = any, IsCustomType extends true | false = false> {
  key: string;
  filter: FilterObjectClass<VType, IsCustomType>;
  updateFilter: (newFilterData: FilterObjectClass) => void;
  fieldData: FilterOption<VType, IsCustomType>;
}

export interface IGetFilterComponentsProps {
  filterData: FilterObjectClass[];
  updateFilter: (filterId: FilterId, newFilterData: FilterObjectClass) => void;
  fieldsData: { [key: string]: FilterOption };
  customFiltersComponents: {
    [key: string]: React.FC<IFilterComponentProps<any, boolean>>;
  };
}

export interface IGetFilterComponentProps<VType = any, IsCustomType extends true | false = false> {
  filter: FilterObjectClass<VType, IsCustomType>;
  updateFilter: (newFilterData: FilterObjectClass) => void;
  fieldData: FilterOption<VType, IsCustomType>;
  customFiltersComponents: {
    [key: string]: React.FC<IFilterComponentProps<any, boolean>>;
  };
}

// export interface INestedFilterOption {
//   uid: Uid;
//   label: string;
//   type: EFilterType.embedded;
//   filters: { [k: string]: FilterOption };
// }
