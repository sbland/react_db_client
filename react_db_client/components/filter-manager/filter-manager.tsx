import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import {
  EComparisons,
  FilterObjectClass,
  FilterOption,
} from '@react_db_client/constants.client-types';
import { useAutoHidePanel } from '@react_db_client/hooks.use-auto-hide-panel-hook';

import { FiltersList } from './FiltersList';
import { AddFilterButton } from './add-filter-button';
import { FilterId, TFilterFunc, IFilterComponentProps } from './lib';

export interface IFilterPanelProps {
  filters: FilterObjectClass[];
  addFilter: (filter: FilterObjectClass) => void;
  deleteFilter: (filterId: FilterId) => void;
  updateFilter: (filterId: FilterId, newFilterData: FilterObjectClass) => void;
  clearFilters: () => void;
  updateFieldTarget: (filterId: FilterId, fieldId: string | number) => void;
  updateOperator: (filterId: FilterId, newOperator: EComparisons) => void;
  showPanelOverride?: boolean;
  fieldsData: { [key: string]: FilterOption };
  floating?: boolean;
  autoOpenPanel?: boolean;
  customFilters?: { [key: string]: TFilterFunc };
  customFiltersComponents?: { [key: string]: React.FC<IFilterComponentProps<any, true>> };
}

/**
 * Filter Manager Panel React Component
 *
 * @param {*} {
 *   filterData,
 *   addFilter, (filterData) => {}
 *   deleteFilter, (index) => {}
 *   updateFilter, (index, newData) => {}
 *   clearFilters, () => {}
 *   showPanelOverride,
 *   fieldsData,
 *   floating,
 *   autoOpenPanel,
 * }
 */
export const FilterPanel = ({
  filters: filterData,
  addFilter,
  deleteFilter,
  updateFilter,
  clearFilters,
  updateFieldTarget,
  updateOperator,
  showPanelOverride,
  fieldsData,
  floating,
  autoOpenPanel,
  customFilters,
  customFiltersComponents,
}: IFilterPanelProps) => {
  const menuRef = useRef<HTMLDivElement>(null);
  const [showPanel, setShowPanel] = useAutoHidePanel(menuRef, floating, showPanelOverride);
  // Auto show panel if values change
  useEffect(() => {
    if (filterData && Object.keys(filterData).length > 0) {
      // TODO: Stop if is search string
      if (autoOpenPanel) setShowPanel(true);
    }
  }, [filterData, autoOpenPanel, setShowPanel]);

  const panelClassName = ['filterPanel_panel', floating ? 'floating' : ''].join(' ');
  return (
    <div className="filterManager" data-testid="rdc-filterManger">
      <button
        type="button"
        className="button-one openFiltersButton"
        onClick={() => setShowPanel(!showPanel)}
      >
        Filters
      </button>
      <div
        className={panelClassName}
        ref={menuRef}
        style={{
          display: showPanel ? 'block' : 'none',
        }}
      >
        <FiltersList
          filterData={filterData}
          deleteFilter={(filterIndex) => deleteFilter(filterIndex)}
          updateFilter={(filterIndex, newFilterData) => updateFilter(filterIndex, newFilterData)}
          fieldsData={fieldsData}
          // customFilters={customFilters}
          customFiltersComponents={customFiltersComponents}
          updateFieldTarget={updateFieldTarget}
          updateOperator={updateOperator}
        />
        <button
          type="button"
          className="button-one"
          onClick={() => {
            clearFilters();
            setShowPanel(false);
          }}
        >
          Clear Filters
        </button>
        <AddFilterButton
          fieldsData={fieldsData}
          returnNewFilter={(newFilterObj) => {
            addFilter(newFilterObj);
          }}
          customFilters={customFilters}
        />
      </div>
    </div>
  );
};

FilterPanel.propTypes = {
  filters: PropTypes.arrayOf(PropTypes.instanceOf(FilterObjectClass)).isRequired,
  addFilter: PropTypes.func.isRequired,
  deleteFilter: PropTypes.func.isRequired,
  updateFilter: PropTypes.func.isRequired,
  clearFilters: PropTypes.func.isRequired,
  showPanelOverride: PropTypes.bool,
  fieldsData: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
    })
  ).isRequired,
  floating: PropTypes.bool,
  autoOpenPanel: PropTypes.bool,
  customFilters: PropTypes.objectOf(PropTypes.func),
  customFiltersComponents: PropTypes.objectOf(PropTypes.elementType),
};

FilterPanel.defaultProps = {
  showPanelOverride: false,
  floating: false,
  autoOpenPanel: false,
  customFilters: {},
  customFiltersComponents: {},
};

export default FilterPanel;
