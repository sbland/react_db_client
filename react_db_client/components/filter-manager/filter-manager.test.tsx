import React from 'react';
import { render, screen, within, waitFor } from '@testing-library/react';
import UserEvent from '@testing-library/user-event';
import cloneDeep from 'lodash/cloneDeep';

import * as compositions from './filter-manager.composition';
import { comparisons, EComparisons } from '@react_db_client/constants.client-types';
import { FilterObjectClass } from '@react_db_client/constants.client-types';
import { FilterPanel, IFilterPanelProps } from './filter-manager';
import {
  demoFiltersData,
  demoFieldsData,
  demoFilterCustom,
  customFilters,
  customFiltersComponents,
  demoFilterString,
  demoFilterNumber,
  customField,
} from './demoData';

const addFilter = jest.fn();
const deleteFilter = jest.fn();
const updateFilter = jest.fn();
const clearFilters = jest.fn();
const updateFieldTarget = jest.fn();
const updateOperator = jest.fn();

const defaultProps: IFilterPanelProps = {
  filters: demoFiltersData,
  addFilter,
  deleteFilter,
  updateFilter,
  clearFilters,
  updateFieldTarget,
  updateOperator,
  showPanelOverride: true,
  fieldsData: demoFieldsData,
  floating: false,
  autoOpenPanel: false,
  customFilters,
  customFiltersComponents,
};

let demoFiltersCopy = cloneDeep(demoFiltersData);

// const demoFilterString = demoFiltersData.find(
//   (f) => f.type === EFilterType.text
// ) as FilterObjectClass;

// const demoFilterNumber = demoFiltersData.find(
//   (f) => f.type === EFilterType.number
// ) as FilterObjectClass;

describe('Filter Manager', () => {
  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        render(<Composition />);
      });
      // test(`Matches Snapshot ${name}`, () => {
      //   render(<Composition />);

      //   const tree = screen.debug();
      //   expect(tree).toMatchSnapshot();
      // });
    });
  });
  describe('Unit Tests', () => {
    beforeEach(() => {
      updateFilter.mockClear();
      deleteFilter.mockClear();
      addFilter.mockClear();
      clearFilters.mockClear();
      updateFieldTarget.mockClear();
      updateOperator.mockClear();
      render(<FilterPanel {...defaultProps} />);
    });
    test('should show a filter for each filter item', () => {
      const filterCount = demoFiltersCopy.length;
      const filterListItems = screen.getAllByRole('listitem');
      expect(filterListItems.length).toBeGreaterThan(0);
      expect(filterListItems.length).toEqual(filterCount);
    });

    test('Entering text into a string filter will update the value in the filter data', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterStringItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterString?.label} field`) //Select name field
      ) as HTMLElement;
      const filterInput: HTMLInputElement = within(filterStringItem).getByRole('textbox');
      await UserEvent.click(filterInput);
      const newFilterData = cloneDeep(demoFilterString);
      const newVal = 'e';
      await UserEvent.type(filterInput, newVal);
      newFilterData.value = `${newFilterData.value}${newVal}`;
      expect(updateFilter).toHaveBeenCalledWith(0, newFilterData);
    });
    test('should call update filter when modifying number filter data', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterNumberItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterNumber?.label} field`) //Select name field
      ) as HTMLElement;
      const filterInput: HTMLInputElement = within(filterNumberItem).getByRole('spinbutton');
      await UserEvent.click(filterInput);
      await UserEvent.clear(filterInput);
      await waitFor(() => filterInput.value === '');
      const newFilterData = cloneDeep(demoFilterNumber);
      newFilterData.value = 0;
      expect(updateFilter).toHaveBeenCalledWith(1, newFilterData);

      const newVal = '1';
      await UserEvent.type(filterInput, newVal);
      newFilterData.value = 1;
      expect(updateFilter).toHaveBeenCalledWith(1, newFilterData);
    });
    test('should call update filter with changed operator when using operator dropdown', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterNumberItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterNumber?.label} field`) //Select name field
      ) as HTMLElement;
      const operatorSelect: HTMLSelectElement = within(filterNumberItem).getByLabelText(
        `Select ${demoFilterNumber?.label} operator`
      );
      expect(operatorSelect.value).toEqual('>');
      await UserEvent.selectOptions(operatorSelect, comparisons.equals);
      expect(updateOperator).toHaveBeenCalledWith(1, comparisons.equals);
    });
    test('should call updateFieldTarget if filter field is changed', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterNumberItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterNumber?.label} field`) //Select name field
      ) as HTMLElement;
      const fieldSelect: HTMLSelectElement = within(filterNumberItem).getByLabelText(
        `Select ${demoFilterNumber?.label} field`
      );
      expect(fieldSelect.value).toEqual('count');
      await UserEvent.selectOptions(fieldSelect, demoFieldsData.filterB.label);

      expect(updateFieldTarget).toHaveBeenCalledWith(1, demoFieldsData.filterB.uid);
    });
    test('should allow custom field types', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterStringItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterCustom?.label} field`) //Select name field
      ) as HTMLElement;
      const customFilterComponent = within(filterStringItem).getByTestId('customFilter');
      const filterButton: HTMLInputElement = within(customFilterComponent).getByRole('button');
      await UserEvent.click(filterButton);
      const newFilterData = cloneDeep(demoFilterCustom);
      newFilterData.value = 'button clicked';
      expect(updateFilter).toHaveBeenCalledWith(3, newFilterData);
    });
    test('should be able to add a filter', async () => {
      const newFilterBtn = screen.getByRole('button', { name: /Add Filter/ });
      await UserEvent.click(newFilterBtn);
      const newFilter = new FilterObjectClass({
        ...demoFilterString,
        uid: expect.any(String),
        label: 'Name',
        operator: EComparisons.EQUALS,
        value: undefined,
      });
      expect(addFilter).toHaveBeenCalledWith(newFilter);
    });
    test('should be able to change filter field to custom field', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterStringItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterString?.label} field`) //Select name field
      ) as HTMLElement;
      const filterFieldSelect: HTMLInputElement = within(filterStringItem).getByLabelText(
        `Select ${demoFilterString?.label} field`
      );
      await UserEvent.selectOptions(filterFieldSelect, customField.label);
      expect(updateFieldTarget).toHaveBeenCalledWith(0, customField.uid);
    });
    test('should be able to have seperate field and data field', () => {
      // This is useful if we need to pass a different search field id to the field we get the filter data from
    });
  });
  describe('Integration Tests', () => {
    beforeEach(() => {
      render(<compositions.FilterManagerWithText />);
    });
    test('should show a filter for each filter item', () => {
      const filterCount = demoFiltersCopy.length;
      const filterListItems = screen.getAllByRole('listitem');
      expect(filterListItems.length).toBeGreaterThan(0);
      expect(filterListItems.length).toEqual(filterCount);
    });

    test('Entering text into a string filter will update the value in the filter data', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterStringItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterString?.label} field`) //Select name field
      ) as HTMLElement;
      const filterInput: HTMLInputElement = within(filterStringItem).getByRole('textbox');
      await UserEvent.click(filterInput);
      await UserEvent.clear(filterInput);
      const newFilterData = cloneDeep(demoFilterString);
      const newVal = 'abc';
      await UserEvent.type(filterInput, newVal);
      newFilterData.value = `${newFilterData.value}${newVal}`;
      expect(filterInput.value).toEqual(newVal);
    });
    test('should call update filter when modifying number filter data', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterNumberItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterNumber?.label} field`) //Select name field
      ) as HTMLElement;
      const filterInput: HTMLInputElement = within(filterNumberItem).getByRole('spinbutton');
      await UserEvent.click(filterInput);
      await UserEvent.clear(filterInput);
      await waitFor(() => filterInput.value === '');
      const newFilterData = cloneDeep(demoFilterNumber);
      newFilterData.value = 0;
      expect(filterInput.value).toEqual('0');

      const newVal = '123';
      await UserEvent.type(filterInput, newVal);
      newFilterData.value = 1;
      expect(filterInput.value).toEqual(newVal);
    });
    test('should call update filter with changed operator when using operator dropdown', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterNumberItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterNumber?.label} field`) //Select name field
      ) as HTMLElement;
      const operatorSelect: HTMLSelectElement = within(filterNumberItem).getByLabelText(
        `Select ${demoFilterNumber?.label} operator`
      );
      expect(operatorSelect.value).toEqual('>');
      await UserEvent.selectOptions(operatorSelect, comparisons.equals);
      expect(operatorSelect.value).toEqual(comparisons.equals);
    });
    test('should call updateFieldTarget if filter field is changed', async () => {
      expect(screen.queryByDisplayValue(demoFieldsData.filterB.label)).not.toBeInTheDocument();
      const filterListItems = screen.getAllByRole('listitem');
      const filterNumberItem = filterListItems.find(
        (f) => within(f).queryByLabelText(`Select ${demoFilterNumber?.label} field`) //Select name field
      ) as HTMLElement;
      const fieldSelect: HTMLSelectElement = within(filterNumberItem).getByLabelText(
        `Select ${demoFilterNumber?.label} field`
      );
      expect(fieldSelect.value).toEqual('count');
      await UserEvent.selectOptions(fieldSelect, demoFieldsData.filterB.label);
      await waitFor(() => fieldSelect.value === demoFieldsData.filterB.uid);
      await waitFor(() => screen.getByDisplayValue(demoFieldsData.filterB.label));
      const newFilter = await screen.findByDisplayValue(demoFieldsData.filterB.label);
      expect(newFilter).toBeInTheDocument();
    });
    test('should be able to add a filter', async () => {
      const newFilterBtn = screen.getByRole('button', { name: /Add Filter/ });
      await UserEvent.click(newFilterBtn);
      const filterListItems = screen.getAllByRole('listitem');
      expect(filterListItems.length).toEqual(demoFiltersCopy.length + 1);
    });
    test('should be able to change filter field to custom field', async () => {
      const filterListItems = screen.getAllByRole('listitem');
      const filterFieldSelect: HTMLInputElement = within(filterListItems[0]).getByLabelText(
        `Select ${demoFilterString?.label} field`
      );
      await UserEvent.selectOptions(filterFieldSelect, customField.label);
      await within(screen.getAllByRole('listitem')[0]).findByText('Custom Filter Component');
    });
  });
});
