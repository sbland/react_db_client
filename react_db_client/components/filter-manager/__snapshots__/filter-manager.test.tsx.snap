// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Manager Compositions Matches Snapshot AllTypes 1`] = `undefined`;

exports[`Filter Manager Compositions Matches Snapshot withText 1`] = `undefined`;
